### PR TITLE
[API] Moving MaskedSoftmaxCELoss and LabelSmoothing to model API

### PR DESCRIPTION
**Pull request by ThomasDelteil**
## Description ##
see title https://github.com/dmlc/gluon-nlp/issues/703

### Essentials ###
- [x] PR's title starts with a category (e.g. [BUGFIX], [MODEL], [TUTORIAL], [FEATURE], [DOC], etc)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented

### CODE REVIEW

**Comment by szha** [comment]
@ThomasDelteil thanks for the change. Could you fix the problems in CI and introduce tests for the new modules?

**Comment by ThomasDelteil** [comment]
@szha how can I see the logs of the build without giving access to my private repository to the CI app?
Screen Shot 2019-06-07 at 12 49 26 PM

**Comment by szha** [comment]
@ThomasDelteil I changed the oauth scope setting to public_repo. Does it work?

**Comment by ThomasDelteil** [comment]
Thanks @szha it works now!

File: src/gluonnlp/loss/label_smoothing.py [hunk 1]
```diff
@@ -155,4 +124,4 @@ def hybrid_forward(self, F, inputs, units=None): # pylint: disable=arguments-dif
         else:
```
Comment (eric-haibin-lin) [inline src/gluonnlp/loss/label_smoothing.py]: It would be great if the documentation includes the math for LabelSmoothing ^^

**Comment by mli** [comment]
Job PR-754/8 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-754/8/index.html

File: src/gluonnlp/loss/label_smoothing.py [hunk 2]
```diff
@@ -133,7 +100,7 @@ def hybrid_forward(self, F, inputs, units=None): # pylint: disable=arguments-dif
 
         Parameters
         ----------
-        F
+        F: mx.sym or mx.nd
```
Comment (JulianSlzr) [inline src/gluonnlp/loss/label_smoothing.py]: I believe we typically omit `F` in docstrings (e.g., see `activation_regularizer.py`), because the forward pass is invoked without passing `F` explicitly (it is done behind the scenes).

File: tests/unittest/test_loss.py [hunk 3]
```diff
@@ -32,3 +33,33 @@ def testTemporalActivationRegularizationLoss():
...
     tar(*[mx.nd.arange(1000).reshape(10, 10, 10),
           mx.nd.arange(1000).reshape(10, 10, 10)])
+    
+def testMaskedSoftmaxCrossEntropyLoss():
+    loss_fn = nlp.loss.SoftmaxCEMaskedLoss()
+    pred = mx.nd.array([[[0,0,10],[10,0,0]]]) #N,T,C 1,2,3
+    label = mx.nd.array([[2,2]])
+    valid_length = mx.nd.array([1,]) 
+    loss = loss_fn(pred, label, valid_length)
+    assert loss < 0.1, "1st timestep prediction is correct"
```
Comment (JulianSlzr) [inline tests/unittest/test_loss.py]: Should the `assert` message be "...prediction is wrong" here?

File: src/gluonnlp/loss/loss.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+# under the License.
+
+"""Loss functions."""
+
+__all__ = ['SoftmaxCEMaskedLoss']
+
+import numpy as np
+from mxnet.gluon.loss import SoftmaxCELoss
+
+class SoftmaxCEMaskedLoss(SoftmaxCELoss):
```
Comment (JulianSlzr) [inline src/gluonnlp/loss/loss.py]: As we move this to the main API, some things to consider (@szha):
- Should this be an alias to `SoftmaxCrossEntropyMaskedLoss` (similar to [Gluon proper](https://mxnet.incubator.apache.org/api/python/gluon/loss.html#mxnet.gluon.loss.SoftmaxCrossEntropyLoss))?
- Should this be `MaskedSoftmaxCELoss` instead (e.g., [as in AllenNLP](https://allenai.github.io/allennlp-docs/api/allennlp.nn.util.html))? More generally, how should we name Masked variants of things?

Comment (eric-haibin-lin) [reply src/gluonnlp/loss/loss.py]: +1 on MaskedSoftmaxCrossEntropyLoss and its alias MaskedSoftmaxCELoss

**Comment by mli** [comment]
Job PR-754/10 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-754/10/index.html

**Comment by mli** [comment]
Job PR-754/11 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-754/11/index.html

**Comment by mli** [comment]
Job PR-754/12 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-754/12/index.html

**Comment by szha** [comment]
@JulianSlzr @eric-haibin-lin how does it look now?

File: tests/unittest/test_loss.py (hunk 3 above)
Comment (ThomasDelteil) [reply tests/unittest/test_loss.py]: The first prediction is correct hence the small loss

Comment (JulianSlzr) [reply tests/unittest/test_loss.py]: Yes, but the message is printed when the assertion fails

File: tests/unittest/test_loss.py [hunk 5]
```diff
@@ -32,3 +33,33 @@ def testTemporalActivationRegularizationLoss():
...
+    valid_length = mx.nd.array([2,]) 
+    loss = loss_fn(pred, label, valid_length)
+    assert loss > 1, "2nd timestep prediction is wrong, loss goes up"
```
Comment (JulianSlzr) [inline tests/unittest/test_loss.py]: Likewise here, the message is printed when the assertion fails. e.g., instead write "2nd timestep prediction was wrong, but loss did not go up"

File: tests/unittest/test_loss.py (hunk 3 above)
Comment (JulianSlzr) [reply tests/unittest/test_loss.py]: Ah I see both of our confusions. Maybe write "1st timestep prediction is correct, but loss was high"?

**Comment by mli** [comment]
Job PR-754/13 is complete. 
Docs are uploaded to http://gluon-nlp-staging.s3-accelerate.dualstack.amazonaws.com/PR-754/13/index.html

**Comment by eric-haibin-lin** [comment]
nice work. Thanks @ThomasDelteil
